ui/app/components/inference/ToolParametersSection.tsx

```diff
@@ -0,0 +1,278 @@
+              <span className="text-fg-muted text-xs font-medium uppercase">
+                Additional Tools ({additionalTools.length})
+              </span>
+              <AdditionalToolsList tools={additionalTools} />
+            </div>
+          )}
+
+          {/* Provider Tools */}
+          {hasProviderTools && (
+            <div className="flex flex-col gap-2">
+              <span className="text-fg-muted text-xs font-medium uppercase">
+                Provider Tools ({providerTools.length})
+              </span>
+              <ProviderToolsList tools={providerTools} />
+            </div>
+          )}
+        </div>
+      </SnippetContent>
+    </SnippetLayout>
+  );
+}
+
+function ToolChoiceBadge({ toolChoice }: { toolChoice: ToolChoice }) {
+  if (typeof toolChoice === "string") {
+    return <Badge variant="outline">{toolChoice}</Badge>;
+  }
+
+  // toolChoice is { specific: string }
+  return (
+    <Badge variant="outline">
+      specific: <span className="font-mono">{toolChoice.specific}</span>
+    </Badge>
+  );
+}
+
+function AdditionalToolsList({ tools }: { tools: Tool[] }) {
+  if (tools.length === 1) {
+    return <ToolCard tool={tools[0]} />;
+  }
+
+  const tabs = tools.map((tool, index) => ({
+    id: String(index),
+    label: getToolName(tool),
+    content: <ToolCard tool={tool} />,
+  }));
+
+  return <SnippetTabs tabs={tabs} />;
+}
+
+function getToolName(tool: Tool): string {
+  if (tool.type === "function") {
+    return tool.name;
+  } else if (tool.type === "openai_custom") {
+    return tool.name;
+  }
+  return "Unknown Tool";
+}
+
+function ToolCard({ tool }: { tool: Tool }) {
+  if (tool.type === "function") {
+    return <FunctionToolCard tool={tool} />;
+  } else if (tool.type === "openai_custom") {
+    return <OpenAICustomToolCard tool={tool} />;
+  }
+  return null;
+}
+
+type FunctionToolWithType = { type: "function" } & FunctionTool;
+type OpenAICustomToolWithType = { type: "openai_custom" } & OpenAICustomTool;
+
+function FunctionToolCard({ tool }: { tool: FunctionToolWithType }) {
+  return (
+    <div className="flex flex-col gap-2">
+      <div className="flex flex-wrap items-center gap-2">
+        <Badge variant="secondary">function</Badge>
+        <span className="font-mono text-sm font-medium">{tool.name}</span>
+        {tool.strict && (
+          <Badge variant="outline" className="text-xs">
+            strict
+          </Badge>
+        )}
+      </div>
+      {tool.description && (
+        <p className="text-fg-muted text-sm">{tool.description}</p>
+      )}
+      <div className="mt-1">
+        <span className="text-fg-muted mb-1 block text-xs font-medium uppercase">
+          Parameters
+        </span>
+        <CodeEditor
+          allowedLanguages={["json"]}
+          value={JSON.stringify(tool.parameters, null, 2)}
+          readOnly
+        />
+      </div>
+    </div>
+  );
+}
+
+function OpenAICustomToolCard({ tool }: { tool: OpenAICustomToolWithType }) {
```

Comment:
These look pretty much the same - perhaps just extract `tool.format` or `tool.parameter` in the component depending on the type to get rid of the duplication?